ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+		})
+		return err
+	}
+
+	return nil
+}
+
+func submitStateCustomRetriableError(err error) error {
+	retry := true
+	aerr, ok := err.(awserr.Error)
+	if ok {
+		switch aerr.Code() {
+		case ecsmodel.ErrCodeInvalidParameterException:
+			retry = false
+		case ecsmodel.ErrCodeAccessDeniedException:
+			retry = false
+		case ecsmodel.ErrCodeClientException:
+			retry = false
+		}
+	}
+	return apierrors.NewRetriableError(apierrors.NewRetriable(retry), err)
+}
+
+func (client *ecsClient) DiscoverPollEndpoint(containerInstanceArn string) (string, error) {
+	resp, err := client.discoverPollEndpoint(containerInstanceArn)
+	if err != nil {
+		return "", err
+	}
+	if resp.Endpoint == nil {
+		return "", errors.New("no endpoint returned; nil")
+	}
+
+	return aws.StringValue(resp.Endpoint), nil
+}
+
+func (client *ecsClient) DiscoverTelemetryEndpoint(containerInstanceArn string) (string, error) {
+	resp, err := client.discoverPollEndpoint(containerInstanceArn)
+	if err != nil {
+		return "", err
+	}
+	if resp.TelemetryEndpoint == nil {
+		return "", errors.New("no telemetry endpoint returned; nil")
+	}
+
+	return aws.StringValue(resp.TelemetryEndpoint), nil
+}
+
+func (client *ecsClient) DiscoverServiceConnectEndpoint(containerInstanceArn string) (string, error) {
+	resp, err := client.discoverPollEndpoint(containerInstanceArn)
+	if err != nil {
+		return "", err
+	}
+	if resp.ServiceConnectEndpoint == nil {
+		return "", errors.New("no ServiceConnect endpoint returned; nil")
+	}
+
+	return aws.StringValue(resp.ServiceConnectEndpoint), nil
+}
+
+func (client *ecsClient) discoverPollEndpoint(containerInstanceArn string) (*ecsmodel.DiscoverPollEndpointOutput,
+	error) {
+	// Try getting an entry from the cache.
+	cachedEndpoint, expired, found := client.pollEndpointCache.Get(containerInstanceArn)
+	if !expired && found {
+		// Cache hit and not expired. Return the output.
+		if output, ok := cachedEndpoint.(*ecsmodel.DiscoverPollEndpointOutput); ok {
+			logger.Info("Using cached DiscoverPollEndpoint", logger.Fields{
+				field.Endpoint:               aws.StringValue(output.Endpoint),
+				field.TelemetryEndpoint:      aws.StringValue(output.TelemetryEndpoint),
+				field.ServiceConnectEndpoint: aws.StringValue(output.ServiceConnectEndpoint),
+				field.ContainerInstanceARN:   containerInstanceArn,
+			})
+			return output, nil
+		}
+	}
+
+	// Cache miss or expired, invoke the ECS DiscoverPollEndpoint API.
+	logger.Debug("Invoking DiscoverPollEndpoint", logger.Fields{
+		field.ContainerInstanceARN: containerInstanceArn,
+	})
+	output, err := client.standardClient.DiscoverPollEndpoint(&ecsmodel.DiscoverPollEndpointInput{
+		ContainerInstance: &containerInstanceArn,
+		Cluster:           aws.String(client.configAccessor.Cluster()),
+	})
+	if err != nil {
+		// If we got an error calling the API, fallback to an expired cached endpoint if
+		// we have it.
+		if expired {
+			if output, ok := cachedEndpoint.(*ecsmodel.DiscoverPollEndpointOutput); ok {
```

Comment:
If we rely on an expired endpoint the agent may not be able to form a connection with ACS. Is this practice already being done in `agent/`? Wondering why we wouldn't just want to stop the container?

Reply:
This is already being done in agent module, yes. See original introducing pull request here for rationale: https://github.com/aws/amazon-ecs-agent/pull/3109

> The acs/tacs endpoints that ecs agent connects have never changed
before, so the current behavior of calling the API every 20 minutes is
not necessary. Lengthening this cache TTL to 12 hours will reduce load
on the ECS service and reduce the chance of customers being throttled.
>
> Additionally, the LRU cache we were using automatically evicted any
value once it expired. This means that this API call could become a
source of complete ecs agent failure in the event of an LSE, in the
event that agent threw away it's cached endpoint and then failed to get
a new one via DiscoverPollEndpoint.
>
> By keeping expired values in the cache, we can fallback to use the
expired value in the event that the DiscoverPollEndpoint API is failing,
thus keeping agent connected to ACS and TACS.

Reply:
Synced offline, no further concern here.